### ISSUE TITLE
Issue with Multiple Parameter Types for POST/PUT Application/JSON

### ISSUE THREAD

**Issue by jonmchan** [ruby-grape/grape#2065]
Considering:
```
params do
  requires :tag, types: [String, Array[String]], desc: "Tag or tags."
end
post "/" do
 ...
end
```

In order to support how the the browser typically handles passing arrays through x-www-form-urlencoded, the current code expects the request to be passed in as one of the following:

* `tag[]=tag1&tag[]=tag2`
* `tag[]=tag1`

However the following are invalid when we have `types: [String, Array[String]]`:

`tag=tag1`

I read the discussion at https://github.com/ruby-grape/grape/issues/693 and can somewhat justify this as the default behavior for passing data from multi-select boxes and the such, but when we look in the case of `content-type: application/json`, the reasoning breaks down:

Works:
`{ "tag[]": "tag1" }` 
`{ "tag": ["tag1"] }`
`{ "tag": ["tag1", "tag2"] }`

However, the following fails:
`{ "tag": "tag1" }`

My request is that I believe the plain `{ "tag": "tag1" }` should work for application/json. It doesn't make any sense that the type is supposedly supporting a string or an array, but you have to put [] in the key for the application/json context. This behavior seems like it should only happen for urlencoded data, if not altogether allow both the `tag=tag1` and the `tag[]=tag1` forms to always work. 

Thanks for reviewing this. Opinions?

**Comment by dblock** [comment]
Looks like a bug to me.

**Comment by jonmchan** [comment]
I tried to set up a unit test, but I'm not sure if the rack-test can reproduce the issue. Not sure if the code exercising form-urlencoded and application/json is exercised from unit tests.
```
      it 'works with multiple representations of arrays and non-array' do
        subject.params do
          optional :a, types: [String, Array[String]]
        end
        subject.post '/' do
          params[:a].to_s
        end
        
        post '/', a: 'test'
        expect(last_response.status).to eq(201)
        expect(last_response.body).to eq('test')

        post '/', 'a[]': 'test'
        expect(last_response.status).to eq(201)
        expect(last_response.body).to eq('["test"]')

        post '/', a: ['array']
        expect(last_response.status).to eq(201)
        expect(last_response.body).to eq('["array"]')
      end
```

**Comment by dblock** [comment]
Put a repro with an app?

**Comment by jonmchan** [comment]
I tested the issue and further found that it actually behaves as expected. The issue is that my code was expecting an array, but was failing on receiving a string.

I have fixed my code with the following:

`params[:tag] = Array(params[:tag])` which forces a single value array datatype when a single string is passed. Leaving this here as it might be useful for someone else.

Thanks for your time @dblock .
